[Inductor] added aten.geometric_ decomp

min-jean-cho opened the pull request:
Fixes #91671


cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

min-jean-cho committed:
added aten.geometric_ decomp

min-jean-cho commented on torch/_decomp/decompositions.py:
```diff
@@ -1962,6 +1962,10 @@ def uniform(
 
+@register_decomposition(aten.geometric_)
+def geometric(self, p):
+    return self.copy_(torch.floor(torch.log(1 - torch.rand_like(self)) / torch.log(torch.tensor(1 - p))) + 1)
```
Thanks @jgong5, actually `1-U` just transforms `U` in range `[0, 1)` (which is the range of `torch.rand`, `torch.rand_like` ) to `(0, 1]` to avoid `log(0)`.  

`log(1-U)` for `U `in `[0, 1)`
`log(U)` for `U` in `(0, 1]` or `(0, 1)`.

And I think the formula https://pytorch.org/docs/stable/generated/torch.Tensor.geometric_.html here is incorrect.

jgong5 replied:
Sure, I understood log(U) and log(1-U) part but log(p) vs. log(1-p) part is different. You mean the doc is incorrect. Shall we fix that too?

min-jean-cho replied:
Do you mean `log(q) = log(1-p)`? `q = 1 - p`

Yes, I think the doc is incorrect (pretty sure..) https://github.com/pytorch/pytorch/pull/92049

min-jean-cho replied:
Oh, you meant `log(p)` if the doc were correct

lezcano replied:
The formula in the docs is certainly non-standard. See https://en.wikipedia.org/wiki/Geometric_distribution. We should improve the docs with a description of what does this model (in this case, probability until the first miss).

For the formula I'd suggest two things. First: How is it implemented in core? And second, consider using `torch.log1p` for better stability around 0.

min-jean-cho replied:
> The formula in the docs is certainly non-standard. See https://en.wikipedia.org/wiki/Geometric_distribution. We should improve the docs with a description of what does this model (in this case, probability until the first miss). How is it implemented in core?

The [core ](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/TransformationHelper.h#L149-L153)implements a standard geometric distribution, the first one described in the linked wiki.  
`P{X=i} = p * q^(i-1), i >= 1, where q = 1 -p` 
`p: probability of success`
`q: probability of failure`
`P{X=i}: probability of first success on the ith trial`

Note that, the inplace geometric https://pytorch.org/docs/stable/generated/torch.Tensor.geometric_.html models the first one described in the linked wiki while `torch.distributions.geometric.Geometric` https://pytorch.org/docs/stable/distributions.html?highlight=geometric#torch.distributions.geometric.Geometric models the second one described in the linked wiki. 

>  And second, consider using torch.log1p for better stability around 0

Thanks @lezcano !

min-jean-cho replied:
Sorry, had a closer look at what `torch.log1p` does. And I don't this `torch.log1p` is suitable to just add `1+input`.

i.e., `floor( log(2-U) / log(2-p) ) + 1` is certainly not the same as  `floor( log(1-U) / log(1-p) ) + 1`

lezcano replied:
See how it's implemented here:
https://pytorch.org/docs/stable/_modules/torch/distributions/geometric.html#Geometric

min-jean-cho replied:
Thanks @lezcano, done, makes sense.

min-jean-cho committed:
(1) use log1p for numerical stability (2) assert generator is none

min-jean-cho committed:
(1) use log1p for numerical stability (2) assert generator is none

min-jean-cho committed:
minor syntax fix

min-jean-cho committed:
scale uniform for numerical stability

min-jean-cho committed:
lint

min-jean-cho committed:
Merge branch 'master' into minjean/inductor_decompose_aten_geometric

min-jean-cho committed:
resolve conflict

lezcano commented on torch/_decomp/decompositions.py:
```diff
@@ -1903,6 +1903,19 @@ def uniform(
+    return self.copy_(
+        torch.floor(
+            torch.log(torch.clamp(torch.rand_like(self), min=tiny))
+            / torch.log1p(torch.tensor(-p))
```
Actually, if `p` is a number, it's better to do `math.log1p(-p)`.
Also, as in the other function, it may be better to implement the `in-place` version in terms of the out-of-place function.

jgong5 reviewed (approved):
LGTM as long as you address the remaining comments from @lezcano

min-jean-cho committed:
move to (1) outplace wrapper (2) math.log1p for constant

min-jean-cho commented on torch/_refs/__init__.py:
```diff
@@ -5114,6 +5114,19 @@ def bucketize(
+    tiny = torch.finfo(self.dtype).tiny
+    return self.copy_(
+        torch.floor(
+            torch.log(torch.clamp(torch.rand_like(self), min=tiny)) / math.log1p(-p)
```
Actually, I realize `torch.finfo` won't do for integer dtype. So it's probably better to represent `log(1-U)` as `log1p(-U)` similar to as in exponential.

min-jean-cho committed:
(1) log(1-U) as log1p(-U) (2) default type promotion

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -5114,6 +5114,22 @@ def bucketize(
+)
+def geometric(self, p, generator=None):
+    assert generator is None
+    return self.copy_(
```
this is wrong, `aten.geometric` is a non-inplace op, the inplace variant will be added with `_make_inplace`, so there should be no `copy_` here. 
@lezcano should there be a test to flag this kind of wrong behavior?

lezcano replied:
It'd be great to have such a test. Let me see if I can put something of that sort next week. This should be a good complement for https://github.com/pytorch/pytorch/pull/92003

lezcano replied:
Addressed in https://github.com/pytorch/pytorch/pull/92243

min-jean-cho replied:
Oops, thanks for the catch, done.

ngimel commented on torch/_inductor/decomposition.py:
```diff
@@ -522,7 +522,9 @@ def bernoulli_p(self, p=0.5, *, generator=None):
 """
-extra_random_decomps = get_decompositions([aten.native_dropout, aten.uniform_])
+extra_random_decomps = get_decompositions(
+    [aten.native_dropout, aten.geometric_, aten.uniform_]
```
should we also add `aten.geometric`, after functionalization the calls to inplace op should be replaced with outplace variant

min-jean-cho replied:
Done, we should also do the same for other in-place random ops

min-jean-cho committed:
(1) fix outplace op decomp (2) add outplace op

min-jean-cho committed:
remove wrong commit

lezcano reviewed (approved):
Nice!

lezcano commented:
Actually, could you also add the relevant OpInfo for this ref (and the operation itself if it doesn't have one)?

min-jean-cho committed:
(1) added geometric opinfo (2) added dtype, parameter check for decomp ref

min-jean-cho committed:
Merge branch 'master' into minjean/inductor_decompose_aten_geometric

min-jean-cho committed:
geometric opinfo

min-jean-cho committed:
Merge branch 'minjean/inductor_decompose_aten_geometric' of https://github.com/min-jean-cho/pytorch into minjean/inductor_decompose_aten_geometric

min-jean-cho committed:
lint, fix ci

sanchitintel commented:
@pytorchbot rebase -b master

min-jean-cho committed:
Merge branch 'pytorch:master' into minjean/inductor_decompose_aten_geometric

min-jean-cho committed:
fix ci

min-jean-cho committed:
mistake

min-jean-cho committed:
see cuda ci

min-jean-cho committed:
fix ci

min-jean-cho committed:
fix ci

min-jean-cho commented:
[Geometric] Hi all, thank you for the review. I see all tests have passed now - I've made few minor changes since the last approvals, could you help have a final scan ?

min-jean-cho committed:
Merge branch 'master' into minjean/inductor_decompose_aten_geometric

min-jean-cho commented:
I will go ahead and merge this as the minor changes after the last approvals are necessary (same changes as #92047)

min-jean-cho commented:
@pytorchbot merge
